domain/schema/model/sql/0019-application.sql

```diff
@@ -56,27 +56,49 @@ CREATE TABLE application_scale (
     REFERENCES application (uuid)
 );
 
-CREATE TABLE application_endpoint_space (
+CREATE TABLE application_exposed_endpoint_space (
     application_uuid TEXT NOT NULL,
-    space_uuid TEXT,
-    CONSTRAINT fk_application_endpoint_space_application
+    -- NULL relation_endpoint_uuid represents the wildcard endpoint.
+    relation_endpoint_uuid TEXT,
+    space_uuid TEXT NOT NULL,
+    CONSTRAINT fk_application_exposed_endpoint_space_application
     FOREIGN KEY (application_uuid)
     REFERENCES application (uuid),
-    CONSTRAINT fk_application_endpoint_space_space
+    CONSTRAINT fk_application_exposed_endpoint_space_relation
+    FOREIGN KEY (relation_endpoint_uuid)
+    REFERENCES relation_endpoint (uuid),
+    CONSTRAINT fk_application_exposed_endpoint_space_space
     FOREIGN KEY (space_uuid)
     REFERENCES space (uuid),
-    PRIMARY KEY (application_uuid, space_uuid)
+    PRIMARY KEY (application_uuid, relation_endpoint_uuid, space_uuid)
 );
 
-CREATE TABLE application_endpoint_cidr (
+-- There is no FK against the CIDR, because it's currently free-form.
+CREATE TABLE application_exposed_endpoint_cidr (
```

Comment:
Can you add some more comments here to say that an unqualified `expose` call causes entries here with the "0.0.0.0/0" and "::/0" CIDRs?

Reply:
Should that not be up to the service and state to document?